lib/driver-utils.js

```diff
@@ -0,0 +1,210 @@
+import semver from 'semver';
+import _ from 'lodash';
+import { getPackageVersion } from './utils';
+import log from './logger';
+import drivers from '../../drivers.json'; // eslint-disable-line import/no-unresolved
+
+
+// add the version to each driver in the hash
+const DRIVER_MAP = _.toPairs(drivers).reduce(function (drivers, [driver, info]) {
+  info.version = getPackageVersion(info.packageName);
+  drivers[driver] = info;
+  return drivers;
+}, {});
+
+// pull the automation names from the driver hash
+function pullAutomationNames () {
+  return _.values(DRIVER_MAP).reduce(function (automationNames, {automationName}) {
+    if (automationName) {
+      automationNames.push(automationName);
+    }
+    return automationNames;
+  }, ['Appium']);
+}
+const AUTOMATION_NAMES = pullAutomationNames();
+
+function pullPlatformNames () {
+  return _.values(DRIVER_MAP).reduce(function (platforms, driver) {
+    if (driver.platformName) {
+      platforms.push(driver.platformName);
+    }
+    return platforms;
+  }, []);
+}
+const PLATFORM_NAMES = pullPlatformNames();
+
+let helpers = {};
+
+// function for testing purposes only
+helpers.injectDriverMap = function injectDriverMap (driverMap) {
+  // circumvent the const-ness of DRIVER_MAP
+  for (const driver of _.keys(DRIVER_MAP)) {
+    delete DRIVER_MAP[driver];
+  }
+  for (const [driver, info] of _.toPairs(driverMap)) {
+    DRIVER_MAP[driver] = info;
+  }
+
+  // circumvent the const-ness of AUTOMATION_NAMES
+  while (AUTOMATION_NAMES.length) {
+    AUTOMATION_NAMES.pop();
+  }
+  for (const name of pullAutomationNames()) {
+    AUTOMATION_NAMES.push(name);
+  }
+
+  // circumvent the const-ness of PLATFORM_NAMES
+  while (PLATFORM_NAMES.length) {
+    PLATFORM_NAMES.pop();
+  }
+  for (const name of pullPlatformNames()) {
+    PLATFORM_NAMES.push(name);
+  }
+};
+
+helpers.getDriverClass = function getDriverClass (packageName, driverClass) {
+  if (DRIVER_MAP[driverClass].driver) {
+    return DRIVER_MAP[driverClass].driver;
+  }
+  try {
+    log.debug(`Loading driver '${driverClass}' from '${packageName}'`);
+    const driver = require(packageName)[driverClass];
+
+    // cache the drivers
+    DRIVER_MAP[driverClass].driver = driver;
+
+    return driver;
+  } catch (err) {
+    log.errorAndThrow(`Unable to load driver '${driverClass}' from package '${packageName}': ${err.message}`);
+  }
+};
+
+function getDriverFromConstraints (caps, driverInfo) {
+  const getAlternativeDriver = function getAlternativeDriver (driverInfo) {
+    let alternativeDriver;
+    for (const constraint of (driverInfo.constraints || [])) {
+      if (constraint.type === 'capability') {
+        if (constraint.capability === 'platformVersion') {
+          const platformVersion = semver.valid(semver.coerce(caps.platformVersion));
```

Comment:
is it by design that `caps` variable is caught in the closure? Shall we use _.memoize instead?